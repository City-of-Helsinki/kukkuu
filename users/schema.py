import graphene
from django.conf import settings
from django.contrib.auth import get_user_model
from django.db import transaction
from graphene import relay
from graphene_django import DjangoConnectionField
from graphene_django.types import DjangoObjectType
from graphql_jwt.decorators import login_required

from common.utils import update_object
from kukkuu.exceptions import KukkuuGraphQLError

from .models import Guardian

User = get_user_model()


LanguageEnum = graphene.Enum(
    "Language", [(l[0].upper(), l[0]) for l in settings.LANGUAGES]
)


class GuardianNode(DjangoObjectType):
    language = LanguageEnum()
    email = graphene.String()

    class Meta:
        model = Guardian
        interfaces = (relay.Node,)

    @classmethod
    def get_queryset(cls, queryset, info):
        return queryset.user_can_view(info.context.user).order_by("last_name")

    def resolve_email(self, info, **kwargs):
        return self.user.email


class UpdateMyProfileMutation(graphene.relay.ClientIDMutation):
    class Input:
        first_name = graphene.String()
        last_name = graphene.String()
        phone_number = graphene.String()
        language = LanguageEnum()

    my_profile = graphene.Field(GuardianNode)

    @classmethod
    @login_required
    @transaction.atomic
    def mutate_and_get_payload(cls, root, info, **kwargs):
        user = info.context.user

        try:
            guardian = user.guardian
        except Guardian.DoesNotExist as e:
            raise KukkuuGraphQLError(e)

        update_object(guardian, kwargs)

        return UpdateMyProfileMutation(my_profile=guardian)


class Query:
    guardians = DjangoConnectionField(GuardianNode)
    my_profile = graphene.Field(GuardianNode)

    @staticmethod
    @login_required
    def resolve_guardians(parent, info, **kwargs):
        return Guardian.objects.user_can_view(info.context.user)

    @staticmethod
    @login_required
    def resolve_my_profile(parent, info, **kwargs):
        return Guardian.objects.filter(user=info.context.user).first()


class Mutation:
    update_my_profile = UpdateMyProfileMutation.Field()
