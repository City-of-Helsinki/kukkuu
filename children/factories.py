import factory

from children.models import Child, Relationship
from projects.models import Project
from users.factories import GuardianFactory


class ChildFactory(factory.django.DjangoModelFactory):
    name = factory.Faker("name")
    birthyear = factory.Faker(
        "random_int", min=2018, max=2023
    )  # NOTE: Many of the tests are freezed to 2020
    postal_code = factory.Faker("postcode", locale="fi_FI")
    project = factory.LazyFunction(lambda: Project.objects.get(year=2020))
    notes = factory.Faker(
        "random_element",
        elements=[
            "",
            "Test notes",
            "Alternative test notes",
            "Longer test notes\nwith multiple lines.",
        ],
    )

    class Meta:
        model = Child


class RelationshipFactory(factory.django.DjangoModelFactory):
    child = factory.SubFactory(ChildFactory)
    guardian = factory.SubFactory(GuardianFactory)
    type = factory.Faker(
        "random_element", elements=[t[0] for t in Relationship.TYPE_CHOICES]
    )

    class Meta:
        model = Relationship


class ChildWithGuardianFactory(ChildFactory):
    relationship = factory.RelatedFactory(RelationshipFactory, "child")
    project = factory.LazyFunction(lambda: Project.objects.get(year=2020))


class ChildWithTwoGuardiansFactory(ChildWithGuardianFactory):
    relationship2 = factory.RelatedFactory(RelationshipFactory, "child")
