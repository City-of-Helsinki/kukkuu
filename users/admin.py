from django.contrib import admin

from children.models import Relationship
from users.models import Guardian


class RelationshipInline(admin.TabularInline):
    model = Relationship
    extra = 0
    fields = ("child", "type", "created_at")
    readonly_fields = ("created_at",)

    def has_change_permission(self, request, obj=None):
        return False


@admin.register(Guardian)
class GuardianAdmin(admin.ModelAdmin):
    inlines = (RelationshipInline,)
