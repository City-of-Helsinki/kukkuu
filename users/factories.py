import factory
from django.contrib.auth import get_user_model
from django.db.models import Q

from languages.models import Language
from users.models import Guardian


class UserFactory(factory.django.DjangoModelFactory):
    first_name = factory.Faker("first_name")
    last_name = factory.Faker("last_name")
    username = factory.Faker("user_name")
    email = factory.Faker("email")

    class Meta:
        model = get_user_model()


class GuardianFactory(factory.django.DjangoModelFactory):
    user = factory.SubFactory(UserFactory)
    first_name = factory.Faker("first_name")
    last_name = factory.Faker("last_name")
    phone_number = factory.Faker("phone_number")

    class Meta:
        model = Guardian

    @factory.post_generation
    def relationships(self, created, extracted, **kwargs):
        count = kwargs.pop("count", None)
        if count:
            from children.factories import RelationshipFactory

            RelationshipFactory.create_batch(count, guardian=self, **kwargs)

    @factory.post_generation
    def languages_spoken_at_home(self, created, extracted, **kwargs):
        if languages_spoken_at_home := extracted:
            filters = Q(alpha_3_code__in=languages_spoken_at_home)
            if None in languages_spoken_at_home:
                filters |= Q(alpha_3_code=None)
            self.languages_spoken_at_home.set(Language.objects.filter(filters))
