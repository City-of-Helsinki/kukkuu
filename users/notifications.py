from django.utils.translation import gettext_lazy as _
from django_ilmoitin.dummy_context import dummy_context
from django_ilmoitin.registry import notifications

from .factories import GuardianFactory


class NotificationType:
    GUARDIAN_EMAIL_CHANGED = "guardian_email_changed"


notifications.register(
    NotificationType.GUARDIAN_EMAIL_CHANGED, _("guardian email changed")
)

guardian = GuardianFactory.build()

dummy_context.update({NotificationType.GUARDIAN_EMAIL_CHANGED: {"guardian": guardian}})
