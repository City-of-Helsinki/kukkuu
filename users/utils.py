from django.contrib.auth import get_user_model
from django.core.exceptions import ValidationError
from django.core.validators import validate_email
from django_ilmoitin.utils import send_notification

from kukkuu.exceptions import InvalidEmailFormatError, VerificationTokenInvalidError
from users.notifications import NotificationType
from verification_tokens.models import VerificationToken

User = get_user_model()


def send_guardian_email_changed_notification(guardian):
    send_notification(
        guardian.email,
        NotificationType.GUARDIAN_EMAIL_CHANGED,
        context={"guardian": guardian},
        language=guardian.language,
    )


def validate_email_verification_token(user: User, verification_token_key: str):
    verification_token = VerificationToken.objects.filter(
        user=user,
        verification_type=VerificationToken.VERIFICATION_TYPE_EMAIL_VERIFICATION,
        key=verification_token_key,
    ).first()
    if not verification_token or not verification_token.is_valid():
        raise VerificationTokenInvalidError("The verification token is invalid")


def validate_guardian_email(email: str):
    try:
        validate_email(email)
    except ValidationError:
        raise InvalidEmailFormatError("Invalid email format")


def validate_guardian_data(guardian_data: dict):
    if "email" in guardian_data:
        validate_guardian_email(guardian_data["email"])
    return guardian_data
