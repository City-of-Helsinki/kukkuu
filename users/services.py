from typing import Optional, Union

from django.db.models import QuerySet
from django.template import Context, Template
from django_ilmoitin.utils import send_notification

from users.models import Guardian
from users.utils import get_communication_unsubscribe_ui_url


class GuardianEmailManagementNotificationService:
    @staticmethod
    def _send_notification(
        guardian: Guardian, notification_type: str, email: str, **kwargs
    ):
        context = kwargs.get("context", {})
        context["guardian"] = guardian
        context["unsubscribe_url"] = get_communication_unsubscribe_ui_url(
            guardian, guardian.language
        )
        send_notification(
            email, notification_type, context=context, language=guardian.language
        )

    @staticmethod
    def send_email_changed_notification(guardian: Guardian):
        """Send an email notification to guardian about a successful
        email change process. The email is now changed.

        Args:
            guardian (Guardian): the recipient guardian
        """
        from users.notifications import NotificationType

        GuardianEmailManagementNotificationService._send_notification(
            guardian, NotificationType.GUARDIAN_EMAIL_CHANGED, guardian.email
        )

    @staticmethod
    def send_email_update_token_notification(
        guardian: Guardian, email: str, verification_token_key: str
    ):
        """Send email update that contains the verification token key,
        that must be used to verify the new email during the change process.

        Args:
            guardian (Guardian): guardian whose email is being changed
            email (str): the new email address where the token is being sent
            verification_token_key (str): token to verify new email
        """
        from users.notifications import NotificationType

        GuardianEmailManagementNotificationService._send_notification(
            guardian,
            NotificationType.GUARDIAN_EMAIL_CHANGE_TOKEN,
            email,
            context={
                "verification_token": verification_token_key,
            },
        )


class AuthServiceNotificationService:
    @staticmethod
    def _send_auth_service_is_changing_notification(
        guardian: Guardian, date_of_change_str: Optional[str] = None
    ):
        """Send the auth service is changing notification to a guardian.

        Args:
            guardian (Guardian): the guardian instance
            date_of_change_str (Optional[str], optional): Date or datetime
                in string format. Defaults to None.
                The actual default should be given in the notification template.
        """
        from users.notifications import NotificationType

        send_notification(
            guardian.email,
            NotificationType.USER_AUTH_SERVICE_IS_CHANGING,
            context={
                "guardian": guardian,
                "date_of_change_str": date_of_change_str,
                "children_event_history_markdown": (
                    AuthServiceNotificationService.generate_children_event_history_markdown(  # noqa
                        guardian
                    )
                ),
            },
            language=guardian.language,
        )

    @staticmethod
    def send_user_auth_service_is_changing_notifications(
        guardians: Optional[Union[QuerySet, list[Guardian]]] = None,
        date_of_change_str: Optional[str] = None,
    ):
        """Send user authentication service is changing notifications
        to guariands as recipients.

        If the guardian list is not explicitly given
        the queryset result of
        `Guardian.objects.for_auth_service_is_changing_notification()`
        will be used as a default.

        Args:
            guardians (Optional[Union[QuerySet, list[Guardian]]], optional):
                explicit list of guardians as recipients. Defaults to None.
            date_of_change_str (Optional[str], optional): Date or datetime
                in string format. Defaults to None.
                The actual default should be given in the notification template.
        """
        if guardians is None:
            guardians = Guardian.objects.for_auth_service_is_changing_notification()

        for guardian in guardians:
            AuthServiceNotificationService._send_auth_service_is_changing_notification(
                guardian, date_of_change_str
            )

    @staticmethod
    def generate_children_event_history_markdown(guardian: Guardian):
        """Generates a Markdown string listing a guardian's children's enrolments."""

        # NOTE: This is markdown, so the line changes and white spaces are important!
        template_string = """
{% for child in guardian.children.all %}# {{ child.name }}
{% for enrolment in child.enrolments.all|dictsort:"occurrence.time" %}1. **{{ enrolment.occurrence.event.name }}:** {{ enrolment.occurrence.time|date:"j.n.Y H:i" }}{% if enrolment.occurrence.event.short_description %}
{{enrolment.occurrence.event.short_description}}{% endif %}
{% endfor %}
{% endfor %}
"""  # noqa

        template = Template(template_string)
        context = Context({"guardian": guardian})
        return template.render(context).strip()
