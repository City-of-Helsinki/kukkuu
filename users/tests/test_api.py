from copy import deepcopy

import pytest
from django.utils import timezone
from guardian.shortcuts import assign_perm

from children.factories import RelationshipFactory
from children.tests.test_api import (
    assert_guardian_matches_data,
    assert_permission_denied,
)
from common.tests.conftest import create_api_client_with_user
from common.tests.utils import assert_match_error_code
from common.utils import get_global_id
from kukkuu.consts import INVALID_EMAIL_FORMAT_ERROR, VERIFICATION_TOKEN_INVALID_ERROR
from projects.factories import ProjectFactory
from users.factories import GuardianFactory
from users.models import Guardian
from users.tests.mutations import (
    REQUEST_EMAIL_CHANGE_TOKEN_MUTATION,
    UPDATE_MY_EMAIL_MUTATION,
    UPDATE_MY_PROFILE_MUTATION,
)
from users.tests.queries import (
    GUARDIANS_QUERY,
    MY_ADMIN_PROFILE_QUERY,
    MY_PROFILE_QUERY,
)
from verification_tokens.factories import UserEmailVerificationTokenFactory


@pytest.fixture(autouse=True)
def autouse_db(db):
    pass


def test_guardians_query_unauthenticated(api_client):
    executed = api_client.execute(GUARDIANS_QUERY)

    assert_permission_denied(executed)


def test_guardians_query_normal_user(snapshot, user_api_client, project):
    GuardianFactory(relationships__count=1, relationships__child__project=project)
    GuardianFactory(
        user=user_api_client.user,
        relationships__count=1,
        relationships__child__project=project,
    )

    executed = user_api_client.execute(GUARDIANS_QUERY)

    snapshot.assert_match(executed)


def test_guardians_query_project_user(
    snapshot, project_user_api_client, project, another_project
):
    guardian_1 = GuardianFactory(
        first_name="Guardian having children in own and another project",
        last_name="Should be visible 1/2",
        relationships__count=1,
        relationships__child__project=project,
    )
    RelationshipFactory(
        guardian=guardian_1,
        child__name="Second child from another project - Should NOT be visible",
        child__project=another_project,
    )

    GuardianFactory(
        first_name="Another project own guardian",
        last_name="Should be visible 2/2",
        user=project_user_api_client.user,
        relationships__count=1,
        relationships__child__project=another_project,
    )

    GuardianFactory(
        first_name="Another project guardian",
        last_name="Should NOT be visible",
        relationships__count=1,
        relationships__child__project=another_project,
    )

    executed = project_user_api_client.execute(GUARDIANS_QUERY)

    snapshot.assert_match(executed)


def test_my_profile_query_unauthenticated(api_client):
    GuardianFactory()

    executed = api_client.execute(MY_PROFILE_QUERY)

    assert executed["data"]["myProfile"] is None
    assert_permission_denied(executed)


def test_my_profile_query(snapshot, user_api_client, project):
    GuardianFactory()
    GuardianFactory(
        user=user_api_client.user,
        relationships__count=1,
        relationships__child__project=project,
    )
    GuardianFactory(relationships__count=1, relationships__child__project=project)

    executed = user_api_client.execute(MY_PROFILE_QUERY)

    snapshot.assert_match(executed)


@pytest.mark.parametrize("guardian_email", ["guardian@example.com", ""])
def test_my_profile_query_email(snapshot, guardian_email):
    guardian = GuardianFactory(email=guardian_email, user__email="user@example.com")
    api_client = create_api_client_with_user(guardian.user)

    executed = api_client.execute(
        """
query MyProfile {
  myProfile {
    email
  }
}
"""
    )

    snapshot.assert_match(executed)


def test_my_profile_no_profile(snapshot, user_api_client):
    GuardianFactory()

    executed = user_api_client.execute(MY_PROFILE_QUERY)

    snapshot.assert_match(executed)


UPDATE_MY_PROFILE_VARIABLES = {
    "input": {
        "firstName": "Updated First Name",
        "lastName": "Updated Last Name",
        "phoneNumber": "Updated phone number",
        "language": "EN",
        "languagesSpokenAtHome": [],
    }
}


def test_update_my_profile_mutation_unauthenticated(api_client):
    executed = api_client.execute(
        UPDATE_MY_PROFILE_MUTATION, variables=UPDATE_MY_PROFILE_VARIABLES
    )

    assert_permission_denied(executed)


def test_update_my_profile_mutation(snapshot, user_api_client, languages):
    GuardianFactory(user=user_api_client.user, language="fi")
    variables = deepcopy(UPDATE_MY_PROFILE_VARIABLES)
    variables["input"]["languagesSpokenAtHome"] = [
        get_global_id(language) for language in languages[0:2]
    ]  # fin, swe

    executed = user_api_client.execute(UPDATE_MY_PROFILE_MUTATION, variables=variables)

    snapshot.assert_match(executed)
    guardian = Guardian.objects.get(user=user_api_client.user)
    assert_guardian_matches_data(guardian, variables["input"])


@pytest.mark.parametrize(
    "new_email, is_valid",
    [
        ("changed-email@kummilapset.fi", True),
        ("INVALID_EMAIL", False),
        ("", False),
        (None, False),
    ],
)
def test_update_my_email_mutation(snapshot, user_api_client, new_email, is_valid):
    initial_email = "initial-email@kummilapset.fi"
    user = user_api_client.user
    GuardianFactory(user=user, email=initial_email)
    verification_token = UserEmailVerificationTokenFactory(user=user, email=new_email)
    variables = {
        "input": {"email": new_email, "verificationToken": verification_token.key}
    }
    executed = user_api_client.execute(UPDATE_MY_EMAIL_MUTATION, variables=variables)
    guardian = Guardian.objects.get(user=user)
    if is_valid:
        snapshot.assert_match(executed)
        assert guardian.email == new_email
    elif new_email is None:
        assert len(executed["errors"]) == 1
        assert 'Variable "$input" got invalid value' in executed["errors"][0]["message"]
    else:
        assert_match_error_code(executed, INVALID_EMAIL_FORMAT_ERROR)
        assert guardian.email == initial_email


def test_update_my_email_mutation_with_invalid_token(user_api_client):
    user = user_api_client.user
    initial_email = "initial-email@kummilapset.fi"
    new_email = "changed-email@kummilapset.fi"
    GuardianFactory(user=user, email=initial_email)
    UserEmailVerificationTokenFactory(user=user, email=new_email)
    variables = {
        "input": {
            "email": new_email,
            "verificationToken": "invalid-key",
        }
    }
    executed = user_api_client.execute(UPDATE_MY_EMAIL_MUTATION, variables=variables)
    assert_match_error_code(executed, VERIFICATION_TOKEN_INVALID_ERROR)
    guardian = Guardian.objects.get(user=user)
    assert guardian.email == initial_email


def test_update_my_email_mutation_with_expired_token(user_api_client, settings):
    settings.VERIFICATION_TOKEN_VALID_MINUTES = 10
    user = user_api_client.user
    initial_email = "initial-email@kummilapset.fi"
    GuardianFactory(user=user, email=initial_email)
    date_expired = timezone.now() - timezone.timedelta(
        minutes=(settings.VERIFICATION_TOKEN_VALID_MINUTES + 1)
    )
    assert date_expired < timezone.now()
    verification_token = UserEmailVerificationTokenFactory(
        user=user,
        expiry_date=date_expired,
    )
    variables = {
        "input": {
            "email": "changed-email@kummilapset.fi",
            "verificationToken": verification_token.key,
        }
    }
    executed = user_api_client.execute(UPDATE_MY_EMAIL_MUTATION, variables=variables)
    assert_match_error_code(executed, VERIFICATION_TOKEN_INVALID_ERROR)
    guardian = Guardian.objects.get(user=user)
    assert guardian.email == initial_email


def test_update_my_email_mutation_with_wrong_email(user_api_client):
    user = user_api_client.user
    initial_email = "initial-email@kummilapset.fi"
    new_email = "changed-email@kummilapset.fi"
    GuardianFactory(user=user, email=initial_email)

    verification_token_with_old_email = UserEmailVerificationTokenFactory(
        user=user, email=user.guardian.email
    )
    variables = {
        "input": {
            "email": new_email,
            "verificationToken": verification_token_with_old_email.key,
        }
    }
    executed = user_api_client.execute(UPDATE_MY_EMAIL_MUTATION, variables=variables)
    assert_match_error_code(executed, VERIFICATION_TOKEN_INVALID_ERROR)

    verification_token_with_old_email = UserEmailVerificationTokenFactory(user=user)
    variables = {
        "input": {
            "email": new_email,
            "verificationToken": verification_token_with_old_email.key,
        }
    }
    executed = user_api_client.execute(UPDATE_MY_EMAIL_MUTATION, variables=variables)
    assert_match_error_code(executed, VERIFICATION_TOKEN_INVALID_ERROR)

    guardian = Guardian.objects.get(user=user)
    assert guardian.email == initial_email


def test_update_my_email_mutation_unauthenticated(api_client):
    variables = {
        "input": {
            "email": "changed-email@kummilapset.fi",
            "verificationToken": "something",
        }
    }
    executed = api_client.execute(UPDATE_MY_EMAIL_MUTATION, variables=variables)
    assert_permission_denied(executed)


def test_request_email_change_token_mutation(snapshot, user_api_client):
    GuardianFactory(user=user_api_client.user)
    variables = {"input": {"email": "new-email@kummilapset.fi"}}
    executed = user_api_client.execute(
        REQUEST_EMAIL_CHANGE_TOKEN_MUTATION, variables=variables
    )
    snapshot.assert_match(executed)


def test_request_email_change_token_mutation_without_guardian(user_api_client):
    variables = {"input": {"email": "new-email@kummilapset.fi"}}
    executed = user_api_client.execute(
        REQUEST_EMAIL_CHANGE_TOKEN_MUTATION, variables=variables
    )
    assert len(executed["errors"]) == 1
    assert "User has no guardian." in executed["errors"][0]["message"]


def test_request_email_change_token_mutation_unauthenticated(api_client):
    variables = {"input": {"email": "new-email@kummilapset.fi"}}
    executed = api_client.execute(
        REQUEST_EMAIL_CHANGE_TOKEN_MUTATION, variables=variables
    )
    assert_permission_denied(executed)


def test_my_admin_profile_unauthenticated(api_client):
    executed = api_client.execute(MY_ADMIN_PROFILE_QUERY)
    assert_permission_denied(executed)


def test_my_admin_profile_normal_user(user_api_client):
    ProjectFactory(year=2021, name="some project")
    executed = user_api_client.execute(MY_ADMIN_PROFILE_QUERY)
    assert executed["data"]["myAdminProfile"]["projects"]["edges"] == []


@pytest.mark.parametrize(
    "has_also_model_perms", (False, True), ids=("no_model_perm", "has_also_model_perm")
)
def test_my_admin_profile_project_admin(
    snapshot, user_api_client, has_also_model_perms
):
    project_1 = ProjectFactory(
        year=2021, name="project where base admin object perm but no other object perms"
    )
    project_2 = ProjectFactory(
        year=2022, name="project where base admin object perm and other object perms"
    )
    assign_perm("admin", user_api_client.user, [project_1, project_2])
    assign_perm("publish", user_api_client.user, project_2)
    assign_perm("manage_event_groups", user_api_client.user, project_2)

    if has_also_model_perms:
        assign_perm("projects.admin", user_api_client.user)
        assign_perm("projects.publish", user_api_client.user)
        assign_perm("projects.manage_event_groups", user_api_client.user)

    ProjectFactory(year=2030, name="project where no object perms")

    executed = user_api_client.execute(MY_ADMIN_PROFILE_QUERY)

    snapshot.assert_match(executed)
