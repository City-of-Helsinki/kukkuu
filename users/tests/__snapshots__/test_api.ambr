# serializer version: 1
# name: test_guardians_query_project_user_with_global_view_families_perm
  dict({
    'data': dict({
      'guardians': dict({
        'edges': list([
          dict({
            'node': dict({
              'email': 'debbie77@example.com',
              'firstName': 'Guardian having children in own and another project',
              'lastName': 'Should be visible 1/2',
              'phoneNumber': '(971)824-0675x0649',
              'relationships': dict({
                'edges': list([
                  dict({
                    'node': dict({
                      'child': dict({
                        'birthyear': 2019,
                        'name': 'Cynthia Holmes',
                        'project': dict({
                          'year': 2020,
                        }),
                      }),
                      'type': 'PARENT',
                    }),
                  }),
                ]),
              }),
            }),
          }),
          dict({
            'node': dict({
              'email': 'michellewalker@example.net',
              'firstName': 'Another project own guardian',
              'lastName': 'Should be visible 2/2',
              'phoneNumber': '420.928.1307',
              'relationships': dict({
                'edges': list([
                  dict({
                    'node': dict({
                      'child': dict({
                        'birthyear': 2022,
                        'name': 'Sara Johnson',
                        'project': dict({
                          'year': 2030,
                        }),
                      }),
                      'type': 'OTHER_RELATION',
                    }),
                  }),
                ]),
              }),
            }),
          }),
        ]),
      }),
    }),
  })
# ---
# name: test_my_admin_profile_project_admin[has_also_model_perm]
  dict({
    'data': dict({
      'myAdminProfile': dict({
        'projects': dict({
          'edges': list([
            dict({
              'node': dict({
                'myPermissions': dict({
                  'canSendToAllInProject': True,
                  'manageEventGroups': True,
                  'publish': True,
                }),
                'name': 'Helsingin kaupunginorkesteri',
              }),
            }),
            dict({
              'node': dict({
                'myPermissions': dict({
                  'canSendToAllInProject': True,
                  'manageEventGroups': True,
                  'publish': True,
                }),
                'name': 'Helsingin Kaupunginteatteri, Suomen Kansallisteatteri, Svenska Teatern, Teatteri ILMI Ö., Q-teatteri ja Nukketeatteri Sampo',
              }),
            }),
            dict({
              'node': dict({
                'myPermissions': dict({
                  'canSendToAllInProject': True,
                  'manageEventGroups': True,
                  'publish': True,
                }),
                'name': 'Cirko - uuden sirkuksen keskus, Hotelli- ja ravintolamuseo, Suomen valokuvataiteen museo, Tanssin talo, Tanssiteatteri Hurjaruuth ja Teatterimuseo',
              }),
            }),
            dict({
              'node': dict({
                'myPermissions': dict({
                  'canSendToAllInProject': True,
                  'manageEventGroups': True,
                  'publish': True,
                }),
                'name': 'Arkkitehtuurimuseo, Designmuseo, Helsingin kaupunginmuseo, Suomen kansallismuseo, Suomen kulttuuriperintökasvatuksen seura ja Tiedemuseo Liekki',
              }),
            }),
            dict({
              'node': dict({
                'myPermissions': dict({
                  'canSendToAllInProject': True,
                  'manageEventGroups': True,
                  'publish': True,
                }),
                'name': 'Helsingin kirjasto- ja liikuntapalvelut',
              }),
            }),
            dict({
              'node': dict({
                'myPermissions': dict({
                  'canSendToAllInProject': True,
                  'manageEventGroups': True,
                  'publish': True,
                }),
                'name': 'HAM Helsingin taidemuseo, Amos Rex, Sointi Jazz Orchestra, UMO Helsinki Jazz Orchestra ja Kansallisgalleria: Ateneumin taidemuseo, Nykytaiteen museo Kiasma ja Sinebrychoffin taidemuseo',
              }),
            }),
            dict({
              'node': dict({
                'myPermissions': dict({
                  'canSendToAllInProject': True,
                  'manageEventGroups': True,
                  'publish': True,
                }),
                'name': 'project where no object perms',
              }),
            }),
            dict({
              'node': dict({
                'myPermissions': dict({
                  'canSendToAllInProject': True,
                  'manageEventGroups': True,
                  'publish': True,
                }),
                'name': 'project where base admin object perm but no other object perms',
              }),
            }),
            dict({
              'node': dict({
                'myPermissions': dict({
                  'canSendToAllInProject': True,
                  'manageEventGroups': True,
                  'publish': True,
                }),
                'name': 'project where base admin object perm and other object perms',
              }),
            }),
          ]),
        }),
      }),
    }),
  })
# ---
# name: test_my_admin_profile_project_admin[no_model_perm]
  dict({
    'data': dict({
      'myAdminProfile': dict({
        'projects': dict({
          'edges': list([
            dict({
              'node': dict({
                'myPermissions': dict({
                  'canSendToAllInProject': False,
                  'manageEventGroups': False,
                  'publish': False,
                }),
                'name': 'project where base admin object perm but no other object perms',
              }),
            }),
            dict({
              'node': dict({
                'myPermissions': dict({
                  'canSendToAllInProject': True,
                  'manageEventGroups': True,
                  'publish': True,
                }),
                'name': 'project where base admin object perm and other object perms',
              }),
            }),
          ]),
        }),
      }),
    }),
  })
# ---
# name: test_my_communication_subscriptions_query_as_logged_in[False]
  dict({
    'data': dict({
      'myCommunicationSubscriptions': dict({
        'firstName': 'Michael',
        'hasAcceptedCommunication': False,
        'language': 'fi',
        'lastName': 'Patton',
      }),
    }),
  })
# ---
# name: test_my_communication_subscriptions_query_as_logged_in[True]
  dict({
    'data': dict({
      'myCommunicationSubscriptions': dict({
        'firstName': 'Michael',
        'hasAcceptedCommunication': True,
        'language': 'fi',
        'lastName': 'Patton',
      }),
    }),
  })
# ---
# name: test_my_communication_subscriptions_query_with_auth_verification_token
  dict({
    'data': dict({
      'myCommunicationSubscriptions': dict({
        'firstName': 'Michael',
        'hasAcceptedCommunication': True,
        'language': 'fi',
        'lastName': 'Patton',
      }),
    }),
  })
# ---
# name: test_my_profile_no_profile
  dict({
    'data': dict({
      'myProfile': None,
    }),
  })
# ---
# name: test_my_profile_query
  dict({
    'data': dict({
      'myProfile': dict({
        'email': 'michellewalker@example.net',
        'firstName': 'Robert',
        'hasAcceptedCommunication': False,
        'language': 'FI',
        'languagesSpokenAtHome': dict({
          'edges': list([
          ]),
        }),
        'lastName': 'Crane',
        'phoneNumber': '303.746.6972x70117',
        'relationships': dict({
          'edges': list([
            dict({
              'node': dict({
                'child': dict({
                  'birthyear': 2019,
                  'name': 'Ashley Hernandez',
                  'postalCode': '28130',
                }),
                'type': 'OTHER_GUARDIAN',
              }),
            }),
          ]),
        }),
      }),
    }),
  })
# ---
# name: test_my_profile_query_email[]
  dict({
    'data': dict({
      'myProfile': dict({
        'email': 'user@example.com',
      }),
    }),
  })
# ---
# name: test_my_profile_query_email[guardian@example.com]
  dict({
    'data': dict({
      'myProfile': dict({
        'email': 'guardian@example.com',
      }),
    }),
  })
# ---
# name: test_request_email_change_token_mutation
  dict({
    'data': dict({
      'requestEmailUpdateToken': dict({
        'email': 'new-email@kummilapset.fi',
        'emailUpdateTokenRequested': True,
      }),
    }),
  })
# ---
# name: test_update_my_communication_subscriptions_as_logged_in[False]
  dict({
    'data': dict({
      'updateMyCommunicationSubscriptions': dict({
        'guardian': dict({
          'firstName': 'Michael',
          'hasAcceptedCommunication': True,
          'language': 'fi',
          'lastName': 'Patton',
        }),
      }),
    }),
  })
# ---
# name: test_update_my_communication_subscriptions_as_logged_in[True]
  dict({
    'data': dict({
      'updateMyCommunicationSubscriptions': dict({
        'guardian': dict({
          'firstName': 'Michael',
          'hasAcceptedCommunication': False,
          'language': 'fi',
          'lastName': 'Patton',
        }),
      }),
    }),
  })
# ---
# name: test_update_my_communication_subscriptions_returns_errors_without_required_args[variables0]
  dict({
    'data': None,
    'errors': list([
      dict({
        'extensions': dict({
          'code': 'GENERAL_ERROR',
        }),
        'locations': list([
          dict({
            'column': 3,
            'line': 3,
          }),
        ]),
        'message': "Variable '$input' got invalid value {}; Field 'hasAcceptedCommunication' of required type 'Boolean!' was not provided.",
      }),
    ]),
  })
# ---
# name: test_update_my_communication_subscriptions_returns_errors_without_required_args[variables1]
  dict({
    'data': None,
    'errors': list([
      dict({
        'extensions': dict({
          'code': 'GENERAL_ERROR',
        }),
        'locations': list([
          dict({
            'column': 3,
            'line': 3,
          }),
        ]),
        'message': "Variable '$input' of required type 'UpdateMyCommunicationSubscriptionsMutationInput!' was not provided.",
      }),
    ]),
  })
# ---
# name: test_update_my_communication_subscriptions_returns_errors_without_required_args[variables2]
  dict({
    'data': None,
    'errors': list([
      dict({
        'extensions': dict({
          'code': 'GENERAL_ERROR',
        }),
        'locations': list([
          dict({
            'column': 3,
            'line': 3,
          }),
        ]),
        'message': "Variable '$input' got invalid value {'authToken': 'what ever'}; Field 'hasAcceptedCommunication' of required type 'Boolean!' was not provided.",
      }),
    ]),
  })
# ---
# name: test_update_my_communication_subscriptions_with_auth_verification_token[False]
  dict({
    'data': dict({
      'updateMyCommunicationSubscriptions': dict({
        'guardian': dict({
          'firstName': 'Michael',
          'hasAcceptedCommunication': True,
          'language': 'fi',
          'lastName': 'Patton',
        }),
      }),
    }),
  })
# ---
# name: test_update_my_communication_subscriptions_with_auth_verification_token[True]
  dict({
    'data': dict({
      'updateMyCommunicationSubscriptions': dict({
        'guardian': dict({
          'firstName': 'Michael',
          'hasAcceptedCommunication': False,
          'language': 'fi',
          'lastName': 'Patton',
        }),
      }),
    }),
  })
# ---
# name: test_update_my_email_mutation[changed-email@kummilapset.fi-True]
  dict({
    'data': dict({
      'updateMyEmail': dict({
        'myProfile': dict({
          'email': 'changed-email@kummilapset.fi',
        }),
      }),
    }),
  })
# ---
# name: test_update_my_profile_mutation
  dict({
    'data': dict({
      'updateMyProfile': dict({
        'myProfile': dict({
          'firstName': 'Updated First Name',
          'hasAcceptedCommunication': False,
          'language': 'EN',
          'languagesSpokenAtHome': dict({
            'edges': list([
              dict({
                'node': dict({
                  'alpha3Code': 'swe',
                }),
              }),
              dict({
                'node': dict({
                  'alpha3Code': 'fin',
                }),
              }),
            ]),
          }),
          'lastName': 'Updated Last Name',
          'phoneNumber': 'Updated phone number',
        }),
      }),
    }),
  })
# ---
